### DISCUSSION TITLE
Support for .NET Framework

### DISCUSSION THREAD

**Discussion by AArnott** [thomhurst/TUnit#1415] [Ideas]
How hard would it be to add support for .NET Framework?

**Comment by thomhurst** [comment]
I've no plans to support net framework. Existing frameworks will all pretty much have existing test suites and for a lot of teams, probably too much effort to migrate those to a new framework for not enough benefit.

Combine that with the fact I'd have to sacrifice newer APIs, or start branching logic with compiler if statements which makes the code harder to maintain.

**Reply by AArnott** [reply]
> probably too much effort to migrate those to a new framework for not enough benefit.

I agree. I wouldn't _migrate_ a whole test project to TUnit at this point. Although I _have_ migrated MSTest and NUnit tests to xUnit in the past, so... if I like TUnit enough who knows?

Really my interest is for my new libraries. For instance my newest Nerdbank.MessagePack library initially targeted only .NET 8 (which was lovely) and then I added .NET Standard 2.0 support because the fact is that there are plenty of would-be users that are still on .NET Framework and I want to support that. 
Nerdbank.MessagePack focuses on AOT support when targeting .NET 8, so using TUnit to help me test the AOT scenario sounds interesting. Enough that I would migrate my existing tests to TUnit. But I can't do that if TUnit itself doesn't allow the tests to also run on .NET Framework. 
So ironically, lack of support for .NET Framework in TUnit is a TUnit adoption blocker for me in a _new_, .NET 8 focused library.

Just some food for thought.

> Combine that with the fact I'd have to sacrifice newer APIs, or start branching logic with compiler if statements which makes the code harder to maintain.

I hear you on all that. Adding .NET Standard 2.0 support to Nerdbank.MessagePack required a few `#if`'s as well, which was sad to see, but probably worth it in the end. 
Do you have any idea how many APIs you're already using that won't compile on .NET Framework?

If I were willing to try adding a `net472` target to your library to frontload the work, it sounds like you might not want to accept a pull request with it anyway because you don't want to maintain it. Is that right? Would you be more interested if you found it didn't impact your coding very much?

**Reply by thomhurst** [reply]
> So ironically, lack of support for .NET Framework in TUnit is a TUnit adoption blocker for me in a new, .NET 8 focused library.

Interesting. I guess that isn't something I considered.

> Would you be more interested if you found it didn't impact your coding very much?

There's other things to consider as well. 

Currently source generated code assumes the .net 8 compiler + lang version or higher. And so uses syntax like collection initializers.

Also I wanted to keep the framework feeling modern, so it uses generic attributes in a lot of places. They're not supported in .NET framework afaik so it'd be a massive overhaul of code? And then all source generators would need updating too.

**Reply by AArnott** [reply]
> Currently source generated code assumes the .net 8 compiler + lang version or higher. And so uses syntax like collection initializers.

I wouldn't worry about that. I have plenty of source generators that emit for .NET Framework projects but require a newer C# lang version than 7.2. It's trivial (and works great) for netfx projects to use newer SDKs and language versions.

Certain C# syntax requires newer APIs to be defined, and C# will gracefully disallow that if the APIs aren't defined. In many cases, packages like PolySharp can fill in those missing APIs so that even those features will start working on .NET Framework targeted projects. 

>  it uses generic attributes in a lot of places. They're not supported in .NET framework afaik

Generic attributes work just fine on .NET Framework, as they are not a runtime feature. It's a language feature. My Nerdbank.MessagePack library defines (and uses) generic attributes on its .NET Standard 2.0 build just fine.

FWIW, I tried compiling your TUnit.Core project to the net472 target. After adding a couple package references as necessary as well to the project file, I got down to just these errors:

```
    D:\source\TUnit\TUnit.Core\AsyncEvent.cs(8,22): error CS0246: The type or namespace name 'Lock' could not be found (are you missing a using directive or an assembly reference?)
    D:\source\TUnit\TUnit.Core\Interfaces\IEventReceiver.cs(5,25): error CS8701: Target runtime doesn't support default interface implementation.
    D:\source\TUnit\TUnit.Core\Data\GetOnlyDictionary.cs(12,29): error CS0246: The type or namespace name 'Lock' could not be found (are you missing a using directive or an assembly reference?)
    D:\source\TUnit\TUnit.Core\Models\LazyHook.cs(5,22): error CS0246: The type or namespace name 'Lock' could not be found (are you missing a using directive or an assembly reference?)
    D:\source\TUnit\TUnit.Core\Interfaces\ITestStartEventReceiver.cs(7,10): error CS8701: Target runtime doesn't support default interface implementation.
    D:\source\TUnit\TUnit.Core\TestDataContainer.cs(18,29): error CS0246: The type or namespace name 'Lock' could not be found (are you missing a using directive or an assembly reference?)
    D:\source\TUnit\TUnit.Core\RunHelpers.cs(48,74): error CS0117: 'MethodImplOptions' does not contain a definition for 'AggressiveOptimization'
    D:\source\TUnit\TUnit.Core\RunHelpers.cs(62,74): error CS0117: 'MethodImplOptions' does not contain a definition for 'AggressiveOptimization'
    D:\source\TUnit\TUnit.Core\RunHelpers.cs(76,74): error CS0117: 'MethodImplOptions' does not contain a definition for 'AggressiveOptimization'
```

The most concerning of these is:

> CS8701: Target runtime doesn't support default interface implementation.

because that's an API breaking change. However looking at the actual interface, it only has one member, and its default implementation is `=> 0`, which could be put under an `#if NET` section easily enough, putting the onus on implementations to add that default value only when they target .NET Framework.

Now, this is just for TUnit.Core. I haven't tried the rest of your projects. Presumably at least TUnit.Engine needs to be ported as well.

**Reply by viceroypenguin** [reply]
nit: newer lang versions can be used with older code. could easily make a check (i've seen this done before) in the nuget package props/target to error if the compiler version is low; but newer compilers can still be used with netfx targets.

**Reply by AArnott** [reply]
Yup. The source generator itself can also inspect the lang version and either accommodate or error out based on that.

I pushed [my branch that gets TUnit.Core compiling on .NET Framework](https://github.com/thomhurst/TUnit/compare/main...AArnott:netfx). 
It's a bit rough, but it gets the point across. Most of the changes didn't require `#if`

**Reply by thomhurst** [reply]
Ah nice - It's been a while since I've used framework now so I'm falling behind on what it can and can't do with new language features!

Seems a lot of my concerns are not actual problems.

In that case - I'm not against targeting net framework too!

> In many cases, packages like PolySharp can fill in those missing APIs

Yeah I'm using `PolyFill` at the moment I think in my SourceGenerators (because they have to target netstandard) - So happy to bring those into the main libraries too. I'd like to keep the dependency tree the same if possible - So not bring in anything more as far as the user is concerned. So Polyfilled libraries are fine because they don't transitively flow through to the consuming project.

> props/target to error if the compiler version is low

Good idea. I currently use the props to set the lang to "latest" - But this is obviously different for every sdk.

**Reply by AArnott** [reply]
> In that case - I'm not against targeting net framework too!

Yay.
Do you want to run with or replace what I've already done? Are you interested in my contributing more of the port? I'm quite content if you want to do it all yourself. :)

> I'd like to keep the dependency tree the same if possible

Agreed. But .NET brings in System.Text.Json automatically while .NET Framework doesn't, so a PackageReference is required for .NET Framework.
You _can_ simply reference the System.Text.Json package unconditionally and everything is fine, but sooner or later, someone will come along and complain that your dependency tree brings down packages you don't need, and so I've taken to try to filter them to just those frameworks that need them.

**Reply by thomhurst** [reply]
> > In that case - I'm not against targeting net framework too!
> 
> Yay. Do you want to run with or replace what I've already done? Are you interested in my contributing more of the port? I'm quite content if you want to do it all yourself. :)

If you're happy to keep contributing then I'm happy to let you continue :) 

> > I'd like to keep the dependency tree the same if possible
> 
> Agreed. But .NET brings in System.Text.Json automatically while .NET Framework doesn't, so a PackageReference is required for .NET Framework. You _can_ simply reference the System.Text.Json package unconditionally and everything is fine, but sooner or later, someone will come along and complain that your dependency tree brings down packages you don't need, and so I've taken to try to filter them to just those frameworks that need them.

Yep that sounds good to me - If only framework needs the reference, then let's conditionally add it only for framework.


Also, is there much difference in the APIs between netstandard2.0 and net472? Because if not, targeting netstandard might be better as we can also run all the way down to net461(?) I believe

**Reply by AArnott** [reply]
> targeting netstandard might be better as we can also run all the way down to net461(?) I believe

net461 [isn't supported any more](https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework).
Any .NET Framework less than 4.7.2 doesn't _properly_ support .NET Standard 2.0. The .NET team tried to 'backport' ns2.0 to prior .NET Framework versions and learned it didn't go well. So 4.7.2 was the first 'next version' they could support ns2.0 for properly. So generally speaking, folks should target net472 if they want to consume ns2.0 assemblies to avoid issues.

As for targeting netstandard2.0, I agree if it isn't onerous we could add that or do that instead of .NET Framework specifically. I would generally be in favor of that for libraries. For test frameworks it's less interesting IMO because test projects should target each runtime rather than ns2.0 so that the ns2.0 library they are testing can be executed on each runtime. But I have had (rare) occasions for libraries of test methods to target ns2.0 so they could be referenced by more endpoint test projects that _do_ reference each runtime, but have my own library only compile once. It's not a high priority in any way, but it could be done.

> If you're happy to keep contributing then I'm happy to let you continue :)

I'll see what else I can whip up then.

**Reply by AArnott** [reply]
https://github.com/thomhurst/TUnit/pull/1436

**Reply by megla-tlanghorst** [reply]
Is there any demo of how this is working now? I tried to use it in a .NET Framework project, because ideally it'll be .NET Standard soon, bit I only get error, stating that C# 12 is a minimum requirement:
```
1>CSC: Error TUNIT_LANG_001 : TUnit requires C# 12 or higher when using Source Generation.
1>ModuleInitializerAttribute.g.cs(1,1): Error CS8370 : Feature 'file-scoped namespace' is not available in C# 7.3. Please use language version 10.0 or greater.
1>StackTraceHiddenAttribute.g.cs(1,1): Error CS8370 : Feature 'file-scoped namespace' is not available in C# 7.3. Please use language version 10.0 or greater.
1>ObjectDumperTests-4442916a459b4dda978c121595f23d08.Generated.cs(8,1): Error CS8370 : Feature 'file-scoped namespace' is not available in C# 7.3. Please use language version 10.0 or greater.
1>AssemblyLoader.g.cs(3,19): Error CS8370 : Feature 'file types' is not available in C# 7.3. Please use language version 11.0 or greater.
1>DisableReflectionScanner.g.cs(3,19): Error CS8370 : Feature 'file types' is not available in C# 7.3. Please use language version 11.0 or greater.
1>ModuleInitializerAttribute.g.cs(8,52): Error CS8370 : Feature 'primary constructors' is not available in C# 7.3. Please use language version 12.0 or greater.
1>StackTraceHiddenAttribute.g.cs(12,51): Error CS8370 : Feature 'primary constructors' is not available in C# 7.3. Please use language version 12.0 or greater.
1>ObjectDumperTests-4442916a459b4dda978c121595f23d08.Generated.cs(12,20): Error CS8370 : Feature 'file types' is not available in C# 7.3. Please use language version 11.0 or greater.
1>DisableReflectionScanner.g.cs(5,3): Error CS8370 : Feature 'module initializers' is not available in C# 7.3. Please use language version 9.0 or greater.
1>AssemblyLoader.g.cs(5,3): Error CS8370 : Feature 'module initializers' is not available in C# 7.3. Please use language version 9.0 or greater.
1>ObjectDumperTests-4442916a459b4dda978c121595f23d08.Generated.cs(14,3): Error CS8370 : Feature 'module initializers' is not available in C# 7.3. Please use language version 9.0 or greater.
```

I tried to disable anything source generation related, because I definitely don't need it right now:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net481</TargetFramework>
    <OutputType>Exe</OutputType>
    <TUnitReflectionScanner>true</TUnitReflectionScanner>
    <TUnitImplicitUsings>false</TUnitImplicitUsings>
    <TUnitAssertionsImplicitUsings>false</TUnitAssertionsImplicitUsings>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="TUnit" Version="0.25.21" />
  </ItemGroup>
</Project>
```

But it seems like it does nothing (except not generate global usings).

**Reply by viceroypenguin** [reply]
NB: you can use C# 12 with .net framework by adding `<LangVersion>12</LangVersion>`, assuming you are using the compiler from net8 or net9.

**Reply by AArnott** [reply]
>  I tried to use it in a .NET Framework project, because ideally it'll be .NET Standard soon

Why? Which runtime would you expect to execute a test project that targets .net standard? Targeting .net standard in a test project seems pointless. 

If you are testing a netstandard library, you should write your test project to multi-target .net framework and .net so that you are testing your netstandard library on both runtimes.

**Reply by megla-tlanghorst** [reply]
> NB: you can use C# 12 with .net framework by adding `<LangVersion>12</LangVersion>`, assuming you are using the compiler from net8 or net9.

Yeah that seems to do the trick, maybe that should be documented somewhere.

> Why? Which runtime would you expect to execute a test project that targets .net standard? Targeting .net standard in a test project seems pointless.
> 
> If you are testing a netstandard library, you should write your test project to multi-target .net framework and .net so that you are testing your netstandard library on both runtimes.

Once it's .netstandard I can probably get away with using .NET 8. Because as long as tests pass in .NET 8 and it compiles in .netstandard, it's probably fine and that's good enough for me. Might not have written my intent very clearly, it's getting far too late...

Now to fully being able to use it I'll just have to wait until the JUnitXml Logger is ported to MTP...

**Reply by AArnott** [reply]
> as long as tests pass in .NET 8 and it compiles in .netstandard, it's probably fine and that's good enough for me.

That's fine. It's a risk assessment you can make. I personally have run into several cases where .NET Framework tests uniquely fail, due to interop that .NET Framework doesn't support, ConcurrentDictionary rejecting a null EqualityComparer argument, or other subtle differences in behavior.
